### PR TITLE
Add =XOD_USERNAME literal

### LINKED ISSUE #1882: As a xoder I want to use XOD_USERNAME literal so that I can substitute my current username for cloud services
### Rationale

We are implementing XOD Cloud Services and it is required to specify a username in various places. For example, in the feed names. To assist easy project/node sharing, the current logged in user name should be representable as a literal.

### User story

Rather than specifying a hard-coded string `"alice"` I can use a special *literal* `=XOD_USERNAME` and in the final turn it will become `alice`, `bob`, or whoever am I:

- I put a text LCD node
- I bind `=XOD_USERNAME` to `L1`
- I upload to the board
- I see my username on the first line

Note, this is a special kind of literal rather than a search-and-replace token inside a string. The latter would require escape handling and make validation harder.

### Acceptance criteria

- [x] The basic scenario works as described
- [x] If I’m not logged in, an error is shown¹
- [x] “Deploy → Show code for Arduino” does not contain my username as a plain string
- [x] If I try to compile the generated code by hand, I get a compiler error which clearly points to a place telling I should define the actual `XOD_USERNAME` value

¹
```markdown
**Not logged in**
The program uses the `=XOD_USERNAME` literal. You should log in to make it work.
Open the Account sidebar to enter your login/password or sign up.
```

### How to implement

- Introduce the new literal kind
- When transpiling to C++ make it use a `XOD_USERNAME` definition
- Place the actual `XOD_USERNAME` definition into the generated source only during the actual compilation procedure.

### PR DESCRIPTION
**Pull request by brusherru**
It closes #1882

### CODE REVIEW

File: packages/xod-arduino/src/transpiler.js [hunk 1]
```diff
@@ -556,10 +569,17 @@ const transformProjectWithImpls = def(
...
           {
-            config: {
-              XOD_DEBUG: liveness === LIVENESS.DEBUG,
-              XOD_SIMULATION: liveness === LIVENESS.SIMULATION,
-            },
+            config: R.merge(
+              {
+                XOD_DEBUG: liveness === LIVENESS.DEBUG,
+                XOD_SIMULATION: liveness === LIVENESS.SIMULATION,
+                XOD_USERNAME_NEEDED: doesContainSpecialStringLiterals(
```
Comment (nkrkv) [inline packages/xod-arduino/src/transpiler.js]: If a project `doesContainSpecialStringLiterals` it doesn’t mean that the `XOD_USERNAME_NEEDED`. Maybe, it’s just `XOD_PROJECT`.

File: packages/xod-arduino/src/types.js [hunk 2]
```diff
@@ -39,10 +39,22 @@ export const Liveness = EnumType('Liveness', [
   LIVENESS.SIMULATION,
 ]);
 
-export const TConfig = Model('TConfig', {
-  XOD_DEBUG: $.Boolean,
-  XOD_SIMULATION: $.Boolean,
-});
+const xodVariablesM = {
```
Comment (nkrkv) [inline packages/xod-arduino/src/types.js]: Does it really need to be extracted out of `TConfig`? If so, why in `camelCase`, what’s so special in it?

File: packages/xod-arduino/src/transpiler.js [hunk 3]
```diff
@@ -800,7 +820,7 @@ export const getPinsAffectedByErrorRaisers = def(
 );
 
 export const transformProject = def(
-  'transformProject :: Project -> PatchPath -> Liveness -> Either Error TProject',
+  'transformProject :: Project -> PatchPath -> Liveness -> XodVariables -> Either Error TProject',
```
Comment (nkrkv) [inline packages/xod-arduino/src/transpiler.js]:
```suggestion
  'transformProject :: Project -> PatchPath -> TConfig -> Either Error TProject',
```

Keep it simple?

File: packages/xod-client/src/editor/components/inspectorWidgets/pinWidgets/StringPinWidget.jsx [hunk 4]
```diff
@@ -38,10 +77,11 @@ const StringWidget = compose(
     },
   })
 )(props => {
+  const showQuotes = props.focused && props.isStringMode;
   const wrapperClassNames = cls('inspector-input-wrapper', {
-    'string-focused': props.focused,
+    'string-focused': showQuotes,
```
Comment (nkrkv) [inline packages/xod-client/src/editor/components/inspectorWidgets/pinWidgets/StringPinWidget.jsx]:
```suggestion
    'with-fake-quotes': showQuotes,
```

Now `string-focused` becomes misleading a bit

**Comment by nkrkv** [comment]
Fount a bug. Bind `=LOL` to a string input.

Expected: The bound value `=LOL` on the patch board is red
Actual: The bound value is yellow (literals starting with `=` are not necessarily strings)

File: packages/xod-arduino/src/types.js (hunk 2 above)
Comment (brusherru) [reply packages/xod-arduino/src/types.js]: It's a subset of two models: internal `TConfig` and `XodVariables` which is used outside the package (from `xod-client`s).
I can combine `Liveness` and `XodVariables` into another one `???Config` type, but I can't get rid of this subset without code duplication. 🤔

File: packages/xod-arduino/src/transpiler.js (hunk 3 above)
Comment (brusherru) [reply packages/xod-arduino/src/transpiler.js]: `TConfig` is an internal type... See the answer above and let's discuss it there.

File: packages/xod-arduino/platform/configuration.tpl.cpp [hunk 5]
```diff
@@ -15,6 +15,12 @@
 // Uncomment to trace the program runtime in the Serial Monitor
 //#define XOD_DEBUG_ENABLE_TRACE
 
+{{#if XOD_USERNAME}}
+#define XOD_USERNAME "{{XOD_USERNAME}}"
+{{else if XOD_USERNAME_NEEDED}}
+//#define XOD_USERNAME "your_username"
+static_assert(false, "The program uses the `=XOD_USERNAME` literal. Uncomment the definition of the `XOD_USERNAME` variable and change the value to your username and remove this line.");
```
Comment (nkrkv) [inline packages/xod-arduino/platform/configuration.tpl.cpp]:
```suggestion
static_assert(false, "Program uses `=XOD_USERNAME` literal. Uncomment line above, put your username there, and remove this line.");
```

Suggest shorter version: it’s a compiler message which should be as short as possible

File: packages/xod-arduino/src/Transpiler.re [hunk 6]
```diff
@@ -18,30 +18,42 @@ type liveness =
   | Debug
   | Simulation;
 
+type xodVariables = Map.String.t(string);
+
 [@bs.module ".."]
-external _transformProject :
-  (Project.t, string, string) => Either.t(Js.Exn.t, TProject.t) =
+external _transformProject:
+  (Project.t, string, string, 'a) => Either.t(Js.Exn.t, TProject.t) =
```
Comment (nkrkv) [inline packages/xod-arduino/src/Transpiler.re]: Why is it `'a` and not `xodVariables`?

File: packages/xod-arduino/src/Transpiler.rei [hunk 7]
```diff
@@ -20,4 +20,7 @@ type liveness =
   | Debug
   | Simulation;
 
-let transpile: (Project.t, PatchPath.t, liveness) => XResult.t(program);
+type xodVariables = Map.String.t(string);
```
Comment (nkrkv) [inline packages/xod-arduino/src/Transpiler.rei]: Sorry for pointing it so late, just recognized the mental conflict: they are not variables, they are constants (from the program point of view), so: `xodVariables` → `xodGlobals`?

File: packages/xod-client/src/user/messages.js [hunk 8]
```diff
@@ -20,3 +20,9 @@ export const LOG_IN_TO_CONTINUE = {
   // TODO: parametrize action
   note: 'The action requires that you are signed in into your XOD account.',
 };
+
+export const NOT_LOGGED_IN = {
+  title: 'Not logged in',
+  note:
+    'The program uses the `=XOD_USERNAME` literal. You should log in to make it work. Open the Account sidebar to enter your login/password or sign up.',
```
Comment (nkrkv) [inline packages/xod-client/src/user/messages.js]: > Open the Account sidebar to enter your login/password or sign up.

It’s a solution, not note

File: packages/xod-arduino/src/constants.js [hunk 9]
```diff
@@ -5,4 +5,8 @@ export const LIVENESS = {
   SIMULATION: 'SIMULATION',
 };
 
+export const SPECIAL_STRING_LITERAL_REPLACEMENTS = {
+  XOD_USERNAME: 'XOD_USERNAME',
+};
```
Comment (nkrkv) [inline packages/xod-arduino/src/constants.js]: This mapping looks singular. Does it ever have a chance to include not a 1-to-1 mapping?

File: packages/xod-cli/src/listrTasks.js [hunk 10]
```diff
@@ -19,7 +19,8 @@ export const transformTask = (patchName, debug) => ({
     ctx.transform = transformProject(
       ctx.project,
       patchName,
-      debug ? LIVENESS.DEBUG : LIVENESS.NONE
+      debug ? LIVENESS.DEBUG : LIVENESS.NONE,
+      {} // TODO: make possible to pass some xod variables here
```
Comment (nkrkv) [inline packages/xod-cli/src/listrTasks.js]:
```suggestion
      {} // TODO: make possible to pass some globals here
```

?

File: packages/xod-client/src/core/containers/App.jsx [hunk 11]
```diff
@@ -137,12 +143,31 @@ export default class App extends React.Component {
     )(eitherTProject);
   }
 
-  transformProjectForTranspiler(liveness) {
+  getXodVariables() {
```
Comment (nkrkv) [inline packages/xod-client/src/core/containers/App.jsx]:
```suggestion
  getGlobals() {
```

File: packages/xod-arduino/src/transpiler.js [hunk 12]
```diff
@@ -540,14 +540,25 @@ const checkForNativePatchesWithTooManyOutputs = def(
...
 /**
  * Transforms Project into TProject.
  * TProject is an object, that ready to be passed into renderer (handlebars)
  * and it has a ready-to-use values, nothing needed to compute anymore.
  */
 const transformProjectWithImpls = def(
-  'transformProjectWithImpls :: Project -> PatchPath -> Liveness -> Either Error TProject',
-  (project, path, liveness) =>
+  'transformProjectWithImpls :: Project -> PatchPath -> Liveness -> StrMap String -> Either Error TProject',
+  (project, path, liveness, xodVariables) =>
```
Comment (nkrkv) [inline packages/xod-arduino/src/transpiler.js]:
```suggestion
  (project, path, liveness, globals) =>
```

?

File: packages/xod-client-electron/src/view/containers/App.jsx [hunk 13]
```diff
@@ -281,7 +281,8 @@ class App extends client.App {
         : this.props.actions.uploadToArduino();
 
     const eitherTProject = this.transformProjectForTranspiler(
-      debug ? LIVENESS.DEBUG : LIVENESS.NONE
+      debug ? LIVENESS.DEBUG : LIVENESS.NONE,
+      this.getXodVariables()
```
Comment (nkrkv) [inline packages/xod-client-electron/src/view/containers/App.jsx]:
```suggestion
      this.getGlobals()
```

File: packages/xod-client/src/core/containers/App.jsx [hunk 14]
```diff
@@ -137,12 +143,31 @@ export default class App extends React.Component {
...
+    if (this.props.user.isNothing) return Either.Left(NOT_LOGGED_IN);
+
+    return Either.of(
+      catMaybies({
+        XOD_USERNAME: R.chain(maybeProp('username'))(this.props.user),
+      })
+    );
+  }
+
+  transformProjectForTranspiler(liveness, eitherXodVariables) {
```
Comment (nkrkv) [inline packages/xod-client/src/core/containers/App.jsx]:
```suggestion
  transformProjectForTranspiler(liveness, eitherGlobals) {
```

File: packages/xod-client/src/core/containers/App.jsx [hunk 15]
```diff
@@ -137,12 +143,31 @@ export default class App extends React.Component {
...
     try {
       return foldMaybe(
         Either.Left(NO_PATCH_TO_TRANSPILE),
         curPatchPath =>
-          transformProject(this.props.project, curPatchPath, liveness),
+          R.chain(
+            xodVariables =>
```
Comment (nkrkv) [inline packages/xod-client/src/core/containers/App.jsx]:
```suggestion
            globals =>
```

File: packages/xod-cli/src/listrTasks.js [hunk 16]
```diff
@@ -19,7 +19,8 @@ export const transformTask = (patchName, debug) => ({
     ctx.transform = transformProject(
       ctx.project,
       patchName,
-      debug ? LIVENESS.DEBUG : LIVENESS.NONE
+      debug ? LIVENESS.DEBUG : LIVENESS.NONE,
+      {} // TODO: make possible to pass some globals here
```
Comment (evgenykochetkov) [inline packages/xod-cli/src/listrTasks.js]: 👍🏻

**Pull request merged**